### ISSUE TITLE
actix-multipart file fields might not be waking properly

### ISSUE THREAD

**Issue by asonix** [actix/actix-web#1679]
When updating to the latest stable versions of crates, actix-multipart seems to have gotten much slower.

## Expected Behavior
actix-multipart's file fields yield the next bytes sequence when asked for

## Current Behavior
actix-multipart's Field stream waits 5 seconds (sometimes) if it can't be polled again immediately (i.e. something else yields to the reactor between polls to the field stream). This happens in release mode, but more consistently in debug mode.

Interestingly, it is almost exactly a 5 second delay between polls. My logs from testing have timestamps like the following:
```
Sep 13 23:02:51.866 DEBUG field: actix_form_data::upload: 32760
Sep 13 23:02:56.864 DEBUG field: actix_form_data::upload: 11382
...
Sep 13 23:02:56.865 DEBUG field: actix_form_data::upload: 53948
Sep 13 23:03:01.865 DEBUG field: actix_form_data::upload: 32759
...
Sep 13 23:03:06.866 DEBUG field: actix_form_data::upload: 32759
Sep 13 23:03:11.866 DEBUG field: actix_form_data::upload: 32768
...
Sep 13 23:03:11.867 DEBUG field: actix_form_data::upload: 32759
Sep 13 23:03:16.867 DEBUG field: actix_form_data::upload: 32768
...
Sep 13 23:03:16.868 DEBUG field: actix_form_data::upload: 32760
Sep 13 23:03:21.868 DEBUG field: actix_form_data::upload: 22963

```

## Possible Solution
Not sure, sorry

## Steps to Reproduce (for bugs)
An easy way to see this happen is to create a dummy future that counts to 500, yielding at each step:
```rust
struct YieldNow(u16, u16);

impl Future for YieldNow {
    type Output = ();

    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
        if self.0 < self.1 {
            self.0 += 1;
            cx.waker().wake_by_ref();
            Poll::Pending
        } else {
            Poll::Ready(())
        }
    }
}
```

Then put it in your multipart handler:
```rust
    while let Some(res) = field.next().await {
        let bytes = res?;
        YieldNow(0, 500).await;
        tracing::debug!("{}", bytes.len());
    }
```

You may need to tweak the number depending on how powerful your system is. 500 works for me to demonstrate the issue, but 100 is too fast.

## Context
I maintain a library for easily parsing form data in actix web, and noticed this issue when upgrading to the 3.0.1 stable release.

## Your Environment

* Rust Version (I.e, output of `rustc -V`): rustc 1.46.0 (04488afe3 2020-08-24)
* Actix Web Version: 3.0.1
* Actix Multipart Version: 0.3.0
* OS: Ubuntu 18.04 with Linux 5.4.0

**Comment by asonix** [comment]
I've narrowed it down to simply affecting an `actix_web::web::Payload` when polled from inside an `actix_rt::spawn(async move { })`, if the payload is polled from the request task there is no issue

**Comment by asonix** [comment]
Here's a full example demonstrating the issue:
```toml
[dependencies]
actix-rt = "1.1.1"
actix-web = "3.0.1"
futures = "0.3.5"
tokio = "0.2.22"
```
```rust
use actix_web::{
    web::{post, resource, BytesMut, Payload},
    App, Error, HttpResponse, HttpServer,
};
use futures::stream::StreamExt;
use std::{
    future::Future,
    pin::Pin,
    task::{Context, Poll},
    time::Instant,
};
use tokio::sync::oneshot::channel;

struct YieldNow(u32, u32);

impl Future for YieldNow {
    type Output = ();

    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
        if self.0 < self.1 {
            self.0 += 1;
            cx.waker().wake_by_ref();
            Poll::Pending
        } else {
            Poll::Ready(())
        }
    }
}

async fn upload(mut body: Payload) -> Result<HttpResponse, Error> {
    let (tx, rx) = channel();

    actix_rt::spawn(async move {
        let mut bytes_mut = BytesMut::new();

        while let Some(Ok(bytes)) = body.next().await {
            println!("{:?} Before count", Instant::now());
            YieldNow(0, 500).await;
            println!("{:?} After count", Instant::now());
            bytes_mut.extend(&bytes);
        }

        tx.send(bytes_mut.freeze()).unwrap();
    });

    let bytes = rx.await.unwrap();
    println!("{:?} {}", Instant::now(), bytes.len());

    Ok(HttpResponse::Created().finish())
}

#[actix_rt::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    HttpServer::new(move || App::new().service(resource("/upload").route(post().to(upload))))
        .bind("127.0.0.1:8080")?
        .run()
        .await?;

    Ok(())
}
```

**Comment by robjtede** [comment]
I haven't been able to reproduce this (macos) using the code without multipart even using high numbers on the YieldNow. Using 20,000 on YieldNow a 2MB file uploaded with cURL takes 300ms and a 20MB file takes 3s.

<details>
<summary>Complete Logs</summary>

cURL:
```
› time curl -v -F file=@2mb 127.0.0.1:8080/upload
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to 127.0.0.1 (127.0.0.1) port 8080 (#0)
> POST /upload HTTP/1.1
> Host: 127.0.0.1:8080
> User-Agent: curl/7.64.1
> Accept: */*
> Content-Length: 2097347
> Content-Type: multipart/form-data; boundary=------------------------edcdfb45bc9b1819
> Expect: 100-continue
>
< HTTP/1.1 100 Continue
* We are completely uploaded and fine
< HTTP/1.1 201 Created
< content-length: 0
< date: Sun, 01 Nov 2020 22:46:19 GMT
<
* Connection #0 to host 127.0.0.1 left intact
* Closing connection 0
curl -v -F file=@2mb 127.0.0.1:8080/upload  0.00s user 0.01s system 3% cpu 0.337 total
› time curl -v -F file=@2mb 127.0.0.1:8080/upload
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to 127.0.0.1 (127.0.0.1) port 8080 (#0)
> POST /upload HTTP/1.1
> Host: 127.0.0.1:8080
> User-Agent: curl/7.64.1
> Accept: */*
> Content-Length: 2097347
> Content-Type: multipart/form-data; boundary=------------------------171fb013d8ed3c9b
> Expect: 100-continue
>
< HTTP/1.1 100 Continue
* We are completely uploaded and fine
< HTTP/1.1 201 Created
< content-length: 0
< date: Sun, 01 Nov 2020 22:46:22 GMT
<
* Connection #0 to host 127.0.0.1 left intact
* Closing connection 0
curl -v -F file=@2mb 127.0.0.1:8080/upload 
0.00s user 0.01s system 2% cpu 0.345 total
```

Server:
```
Instant { t: 36271612984766 } Before count
Instant { t: 36271614584536 } After count
Instant { t: 36271622223476 } Before count
Instant { t: 36271623542354 } After count
Instant { t: 36271627239054 } Before count
Instant { t: 36271628457627 } After count
Instant { t: 36271632672742 } Before count
Instant { t: 36271633833827 } After count
Instant { t: 36271638070139 } Before count
Instant { t: 36271639412784 } After count
Instant { t: 36271644366824 } Before count
Instant { t: 36271645551131 } After count
Instant { t: 36271650005978 } Before count
Instant { t: 36271651315708 } After count
Instant { t: 36271655322036 } Before count
Instant { t: 36271656474686 } After count
Instant { t: 36271660114608 } Before count
Instant { t: 36271661562424 } After count
Instant { t: 36271665653583 } Before count
Instant { t: 36271666963613 } After count
Instant { t: 36271670782634 } Before count
Instant { t: 36271672019880 } After count
Instant { t: 36271676206277 } Before count
Instant { t: 36271677399612 } After count
Instant { t: 36271680866520 } Before count
Instant { t: 36271682210456 } After count
Instant { t: 36271685931847 } Before count
Instant { t: 36271687051343 } After count
Instant { t: 36271690640366 } Before count
Instant { t: 36271691883875 } After count
Instant { t: 36271695541196 } Before count
Instant { t: 36271696948204 } After count
Instant { t: 36271700753691 } Before count
Instant { t: 36271701978579 } After count
Instant { t: 36271706174284 } Before count
Instant { t: 36271707631212 } After count
Instant { t: 36271711675617 } Before count
Instant { t: 36271713230225 } After count
Instant { t: 36271717108225 } Before count
Instant { t: 36271718196030 } After count
Instant { t: 36271722232791 } Before count
Instant { t: 36271724371538 } After count
Instant { t: 36271728279557 } Before count
Instant { t: 36271729418482 } After count
Instant { t: 36271733517054 } Before count
Instant { t: 36271735313745 } After count
Instant { t: 36271739500861 } Before count
Instant { t: 36271740991809 } After count
Instant { t: 36271744812038 } Before count
Instant { t: 36271746190394 } After count
Instant { t: 36271750394995 } Before count
Instant { t: 36271751963316 } After count
Instant { t: 36271756381025 } Before count
Instant { t: 36271757670461 } After count
Instant { t: 36271761842082 } Before count
Instant { t: 36271763185657 } After count
Instant { t: 36271766764186 } Before count
Instant { t: 36271767855570 } After count
Instant { t: 36271771921066 } Before count
Instant { t: 36271773233022 } After count
Instant { t: 36271777234675 } Before count
Instant { t: 36271778665723 } After count
Instant { t: 36271783262358 } Before count
Instant { t: 36271784355669 } After count
Instant { t: 36271788545554 } Before count
Instant { t: 36271789670859 } After count
Instant { t: 36271793355762 } Before count
Instant { t: 36271794613154 } After count
Instant { t: 36271798309482 } Before count
Instant { t: 36271799549932 } After count
Instant { t: 36271803888647 } Before count
Instant { t: 36271805186144 } After count
Instant { t: 36271808947144 } Before count
Instant { t: 36271810148165 } After count
Instant { t: 36271813970079 } Before count
Instant { t: 36271815319901 } After count
Instant { t: 36271820092981 } Before count
Instant { t: 36271821848818 } After count
Instant { t: 36271825776732 } Before count
Instant { t: 36271827093524 } After count
Instant { t: 36271830810583 } Before count
Instant { t: 36271832420495 } After count
Instant { t: 36271836214995 } Before count
Instant { t: 36271837321907 } After count
Instant { t: 36271841530068 } Before count
Instant { t: 36271842652155 } After count
Instant { t: 36271846473773 } Before count
Instant { t: 36271847730050 } After count
Instant { t: 36271851700998 } Before count
Instant { t: 36271852986681 } After count
Instant { t: 36271856950934 } Before count
Instant { t: 36271858098708 } After count
Instant { t: 36271861810860 } Before count
Instant { t: 36271862908424 } After count
Instant { t: 36271867051498 } Before count
Instant { t: 36271868332045 } After count
Instant { t: 36271872042514 } Before count
Instant { t: 36271873417668 } After count
Instant { t: 36271877037782 } Before count
Instant { t: 36271878322832 } After count
Instant { t: 36271882454957 } Before count
Instant { t: 36271883577754 } After count
Instant { t: 36271887083587 } Before count
Instant { t: 36271888835094 } After count
Instant { t: 36271892306208 } Before count
Instant { t: 36271893970682 } After count
Instant { t: 36271897983187 } Before count
Instant { t: 36271899337437 } After count
Instant { t: 36271903244936 } Before count
Instant { t: 36271904340354 } After count
Instant { t: 36271908668265 } Before count
Instant { t: 36271909703116 } After count
Instant { t: 36271913762265 } Before count
Instant { t: 36271914811768 } After count
Instant { t: 36271918219873 } Before count
Instant { t: 36271919498742 } After count
Instant { t: 36271923241383 } Before count
Instant { t: 36271924319630 } After count
Instant { t: 36271927786245 } Before count
Instant { t: 36271929122185 } After count
Instant { t: 36271932532634 } Before count
Instant { t: 36271933769184 } After count
Instant { t: 36271938261654 } Before count
Instant { t: 36271939294608 } After count
Instant { t: 36271942755218 } Before count
Instant { t: 36271943834985 } After count
Instant { t: 36271947347709 } Before count
Instant { t: 36271948486380 } After count
Instant { t: 36271948597080 } 2097347
```

</details>

**Comment by asonix** [comment]
You're right. I just tried with that curl command uploading 20MB and 200MB and the time it took seemed reasonable and there weren't any gaps. This might then have something to do with the interaction between my test client (async python) and the example server. 

While this seems like it could be an issue on the client side, I don't think it is, since removing the `actix::spawn` from the server side dramatically improves the speed. My hunch is that the server can't handle when data is too slow. Maybe when there is more data to be read immediately, there is no issue (meaning the curl client is too fast to demonstrate the problem).

Here's the client code I've been using to test. I have a file called 20m.file which I created with `fallocate -l 20M 20m.file`

```python
#!/usr/bin/env python3

import asyncio
import aiofiles
import aiohttp

url = 'http://localhost:8080/upload'

async def file_sender(file_name=None):
    async with aiofiles.open(file_name, 'rb') as f:
        chunk = await f.read(64*1024)
        while chunk:
            yield chunk
            chunk = await f.read(64*1024)


async def req():
    async with aiohttp.ClientSession() as session:
        data = aiohttp.FormData(quote_fields=False)
        data.add_field("files[]", file_sender(file_name="20m.file"), filename="20m.file")

        async with session.post(url, data=data) as resp:
            text = await resp.text()
            print(text)
            assert 201 == resp.status


loop = asyncio.get_event_loop()
loop.run_until_complete(req())

```

**Comment by JojiiOfficial** [comment]
I'm experiencing the exact same behavior. They are pretty rare in release builds but they exist. I can't really stream 1GB multipart files without waiting for a long time using a debug build.

**Comment by asonix** [comment]
@JojiiOfficial what kind of client are you using for this?

**Comment by JojiiOfficial** [comment]
I noticed it when rewriting a backend (file upload via REST+Multipart) in rust using actix web. It works perfectly fine with the current backend I'm using (which is written in go). However I got those random upload breaks using the rewritten backend, which only saves the multipart stream into a file.
To answer your question: I'm using a custom (self written) client.

**Comment by asonix** [comment]
I meant my question as more of a "Are you using the same async python libraries I'm using" for the client, because if so, it might have been an issue with the client rather than the server, but it seems that this issue in Actix Web might be real

**Comment by JojiiOfficial** [comment]
With actix v4 (beta) this issue seems to be fixed though.

**Comment by asonix** [comment]
Confirmed with my test program as well, 3.2.2 has this issue, but 4.0.0-beta.4 does not

**Comment by robjtede** [comment]
that's good news, would like to bisect and find the fix commit before closing

**Comment by asonix** [comment]
4.0.0-beta.1 has the issue
4.0.0-beta.2 does not have this issue

**Comment by asonix** [comment]
Looks like it's fixed between 41bc04b1c43c860225b5914d97cf9f7741f1d74d and 7cfed73be8d4349d0e2795e4fde1b3beba5e77c0

PR: https://github.com/actix/actix-web/pull/1929

**Comment by robjtede** [comment]
Makes sense. So it's technically not fixed as @fakeshadow has pointed out on Discord.

**Comment by asonix** [comment]
which channel & when did that conversation happen?

**Comment by fakeshadow** [comment]
The root cause of this problem could be related to this line https://github.com/actix/actix-web/blob/78fcd0237a4033e9802313de0299ce9cbb0a614c/actix-http/src/h1/payload.rs#L240. 

When the payload buffer is exactly at 32kb on first poll it would skip register the waker for future wake up. This could be a perf choice to reduce contention on wakeup but could end up lose the waker when inside a spawn task.

With current master branch the chance of buffer size end up at exactly 32kb is extremely low so I believe for now no further change are needed for this issue unless it shows it still happens in regular manner.

**Comment by thalesfragoso** [comment]
So, this is my theory (using links from beta.1):

During `YieldNow.await`, we fill both the payload and the dispatcher's `read_buf`. We're now backpressuring io (`HW_BUFFER_SIZE` is 32k):
https://github.com/actix/actix-web/blob/c09186a2c07076a02d53137f31fef59f78fe95d5/actix-http/src/h1/dispatcher.rs#L919-L921

Which means that only the payload can now wake up the dispatcher. After `YieldNow.await` we are back polling the payload, which yields a chunk (bigger than 32k in this example), waking the dispatcher, by doing so, it loses its dispatcher waker.

Dispatcher is still with a `read_buf` bigger than 32k, `read_available` returns without registering the waker on the io. It then calls `poll_request`, which doesn't register the waker on the payload because the payload can read immediately:
https://github.com/actix/actix-web/blob/c09186a2c07076a02d53137f31fef59f78fe95d5/actix-http/src/h1/payload.rs#L130-L143

It then pops a chunk (smaller than 32k) from `read_buf`, it tries to pop another one but it's partial, returns.

Next stop, `poll_response`, it polls the service future that returns pending (it's waiting on the whole payload from the spawned task), `poll_response` then calls `poll_request` again, payload still can read more (it has a chunk smaller than 32k), again we don't register our waker in the payload, tries to pop the same chunk again, its partial, returns `false` causing `poll_response` to return with `PollResponse::DoNothing`.

And that's it, we now completely lost the dispatcher's waker, only thing that has it is the keep-alive timer and the service. Payload proceeds to pop its chunk, tries to wake the dispatcher but it doesn't have its waker. Wait for the keep-alive timer, rinse and repeat.

I think this can still happen on the current code, but now we need a `read_buf` bigger than 128k, containing one or more chunks totalling less than 32k and the rest being a partial chunk. Which seems way less probable to happen. A solution could be to insert something like
```rust
if backpressuring_io {
    cx.waker().wake_by_ref();
}
``` 

somewhere in `poll_request`.

**Comment by robjtede** [comment]
> So, this is my theory (using links from beta.1) ...

Following the code path, this explanation seems pretty convincing and the fix suggested in #2369 (applied as 8ff489aa) does not, I believe, tip the scale too far towards over-waking.
